diff --git a/drivers/power/supply/axp20x_battery.c b/drivers/power/supply/axp20x_battery.c
index f6c849bd15e3..eeb70e14fffa 100644
--- a/drivers/power/supply/axp20x_battery.c
+++ b/drivers/power/supply/axp20x_battery.c
@@ -220,6 +220,11 @@ static int axp20x_battery_get_prop(struct power_supply *psy,
 	int ret = 0, reg, val1;
 
 	switch (psp) {
+	case POWER_SUPPLY_PROP_TECHNOLOGY:
+		/* TODO: AXP can be both Li-ion and LiPo */
+		val->intval = POWER_SUPPLY_TECHNOLOGY_LION;
+		break;
+
 	case POWER_SUPPLY_PROP_PRESENT:
 	case POWER_SUPPLY_PROP_ONLINE:
 		ret = regmap_read(axp20x_batt->regmap, AXP20X_PWR_OP_MODE,
