diff --git a/.gitignore b/.gitignore
index 70580bdd352c..a32c8df8fe1a 100644
--- a/.gitignore
+++ b/.gitignore
@@ -67,11 +67,6 @@ modules.order
 #
 /*.spec
 
-#
-# Debian directory (make deb-pkg)
-#
-/debian/
-
 #
 # Snap directory (make snap-pkg)
 #
diff --git a/Makefile b/Makefile
index d4d36c61940b..a9b9d735497c 100644
--- a/Makefile
+++ b/Makefile
@@ -1365,7 +1365,7 @@ CLEAN_FILES += modules.builtin.modinfo
 # Directories & files removed with 'make mrproper'
 MRPROPER_DIRS  += include/config include/generated          \
 		  arch/$(SRCARCH)/include/generated .tmp_objdiff \
-		  debian/ snap/ tar-install/
+		  snap/ tar-install/
 MRPROPER_FILES += .config .config.old .version \
 		  Module.symvers \
 		  signing_key.pem signing_key.priv signing_key.x509	\
diff --git a/scripts/package/builddeb b/scripts/package/builddeb
index 6e8ab87a1305..9f7f495da7aa 100755
--- a/scripts/package/builddeb
+++ b/scripts/package/builddeb
@@ -51,8 +51,8 @@ tmpdir="$objtree/debian/tmp"
 kernel_headers_dir="$objtree/debian/hdrtmp"
 libc_headers_dir="$objtree/debian/headertmp"
 dbg_dir="$objtree/debian/dbgtmp"
-packagename=linux-image-$version
-kernel_headers_packagename=linux-headers-$version
+packagename=linux-image-pine64
+kernel_headers_packagename=linux-headers-pine64
 libc_headers_packagename=linux-libc-dev
 dbg_packagename=$packagename-dbg
 
@@ -143,9 +143,9 @@ if [ "$ARCH" != "um" ]; then
 	$MAKE -f $srctree/Makefile headers_install INSTALL_HDR_PATH="$libc_headers_dir/usr"
 	# move asm headers to /usr/include/<libc-machine>/asm to match the structure
 	# used by Debian-based distros (to support multi-arch)
-	host_arch=$(dpkg-architecture -a$(cat debian/arch) -qDEB_HOST_MULTIARCH)
-	mkdir $libc_headers_dir/usr/include/$host_arch
-	mv $libc_headers_dir/usr/include/asm $libc_headers_dir/usr/include/$host_arch/
+	#host_arch=$(dpkg-architecture -a$(cat debian/arch) -qDEB_HOST_MULTIARCH)
+	#mkdir $libc_headers_dir/usr/include/$host_arch
+	#mv $libc_headers_dir/usr/include/asm $libc_headers_dir/usr/include/$host_arch/
 fi
 
 # Install the maintainer scripts
@@ -165,9 +165,25 @@ set -e
 export DEB_MAINT_PARAMS="\$*"
 
 # Tell initramfs builder whether it's wanted
-export INITRD=$(if_enabled_echo CONFIG_BLK_DEV_INITRD Yes No)
+# export INITRD=$(if_enabled_echo CONFIG_BLK_DEV_INITRD Yes No)
+export INITRD=No
 
 test -d $debhookdir/$script.d && run-parts --arg="$version" --arg="/$installed_image_path" $debhookdir/$script.d
+
+case "$script" in
+postinst)
+	depmod $version
+	rm -f /boot/Image.gz /boot/*.dtb
+	cp -f /boot/Image.gz-$version /boot/Image.gz
+	cp -f /usr/lib/linux-image-pine64/allwinner/sun50i-a64-dontbeevil.dtb /boot
+	cp -f /usr/lib/linux-image-pine64/allwinner/sun50i-a64-pinephone.dtb /boot
+	cp -f /usr/lib/linux-image-pine64/allwinner/sun50i-a64-pinetab.dtb /boot
+	;;
+postrm)
+	rm -f /boot/Image.gz /boot/*.dtb
+	;;
+esac
+
 exit 0
 EOF
 	chmod 755 "$tmpdir/DEBIAN/$script"
@@ -211,7 +227,7 @@ rm -f "$objtree/debian/hdrsrcfiles" "$objtree/debian/hdrobjfiles"
 
 if [ "$ARCH" != "um" ]; then
 	create_package "$kernel_headers_packagename" "$kernel_headers_dir"
-	create_package "$libc_headers_packagename" "$libc_headers_dir"
+	#create_package "$libc_headers_packagename" "$libc_headers_dir"
 fi
 
 create_package "$packagename" "$tmpdir"
