--- a/drivers/media/i2c/ov5640.c	2023-01-10 00:23:34.242742039 +0300
+++ b/drivers/media/i2c/ov5640.c	2023-01-10 00:24:22.825876687 +0300
@@ -3769,7 +3769,7 @@
 	/* Auto/manual gain */
 	ctrls->auto_gain = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_AUTOGAIN,
 					     0, 1, 1, 1);
-	ctrls->gain = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_GAIN,
+	ctrls->gain = v4l2_ctrl_new_std(hdl, ops, V4L2_CID_ANALOGUE_GAIN,
 					0, 1023, 1, 0);
 
 	/* Autofocus */
--- a/drivers/media/platform/sunxi/sun6i-csi/sun6i_video.c	2023-01-10 00:48:37.304517933 +0300
+++ b/drivers/media/platform/sunxi/sun6i-csi/sun6i_video.c	2023-01-10 01:48:27.537006785 +0300
@@ -451,6 +451,39 @@
 	return sun6i_video_format_try(video, format);
 }
 
+static int vidioc_enum_framesizes(struct file *file, void *priv,
+				  struct v4l2_frmsizeenum *f)
+{
+if (f->index != 0)
+		return -EINVAL;
+
+	f->type = V4L2_FRMSIZE_TYPE_CONTINUOUS;
+	f->stepwise.min_width = MIN_WIDTH;
+	f->stepwise.max_width = MAX_WIDTH;
+	f->stepwise.step_width = 1;
+	f->stepwise.min_height = MIN_HEIGHT;
+	f->stepwise.max_height = MAX_HEIGHT;
+	f->stepwise.step_height = 1;
+	return 0;
+}
+
+static int vidioc_enum_frameintervals(struct file *file, void *priv,
+				      struct v4l2_frmivalenum *f)
+{
+	if (f->index > 0)
+		return -EINVAL;
+
+	f->type = V4L2_FRMIVAL_TYPE_CONTINUOUS;
+	f->stepwise.min.denominator = 2;
+	f->stepwise.min.numerator = 1;
+	// Unknown limit
+	f->stepwise.max.denominator = 120;
+	f->stepwise.max.numerator = 1;
+	f->stepwise.step.denominator = 1;
+	f->stepwise.step.numerator = 1;
+	return 0;
+}
+
 static int sun6i_video_enum_input(struct file *file, void *private,
 				  struct v4l2_input *input)
 {
@@ -513,6 +546,8 @@
 	.vidioc_g_fmt_vid_cap		= sun6i_video_g_fmt,
 	.vidioc_s_fmt_vid_cap		= sun6i_video_s_fmt,
 	.vidioc_try_fmt_vid_cap		= sun6i_video_try_fmt,
+	.vidioc_enum_framesizes		= vidioc_enum_framesizes,
+	.vidioc_enum_frameintervals	= vidioc_enum_frameintervals,
 
 	.vidioc_enum_input		= sun6i_video_enum_input,
 	.vidioc_g_input			= sun6i_video_g_input,
@@ -754,7 +789,7 @@
 	/* Video Device */
 
 	strscpy(video_dev->name, SUN6I_CSI_NAME, sizeof(video_dev->name));
-	video_dev->device_caps = V4L2_CAP_VIDEO_CAPTURE | V4L2_CAP_STREAMING;
+	video_dev->device_caps = V4L2_CAP_VIDEO_CAPTURE | V4L2_CAP_STREAMING | V4L2_CAP_IO_MC;
 	video_dev->vfl_dir = VFL_DIR_RX;
 	video_dev->release = video_device_release_empty;
 	video_dev->fops = &sun6i_video_fops;
